src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1017 @@
+		return stripe_end - recx->rx_idx;
+	else
+		return recx->rx_nr;
+}
+
+/* Iterator call back sub-function for handling data extents.
+ */
+static int
+agg_data_extent(vos_iter_entry_t *entry, struct ec_agg_entry *agg_entry,
+		unsigned int *acts)
+{
+	struct ec_agg_extent	*extent = NULL;
+	daos_off_t		 cur_stripenum, this_stripenum;
+	int			 rc = 0;
+
+	D_ASSERT(!(entry->ie_recx.rx_idx & PARITY_INDICATOR));
+
+	this_stripenum = agg_stripenum(agg_entry, entry->ie_recx.rx_idx);
+
+	if (this_stripenum != agg_entry->ae_cur_stripe.as_stripenum) {
+		/* Iterator has reached next stripe */
+		if (agg_entry->ae_cur_stripe.as_extent_cnt) {
+			cur_stripenum = agg_entry->ae_cur_stripe.as_stripenum;
+			rc = agg_process_stripe(agg_entry);
+			if (rc)
+				D_ERROR("Process stripe returned "DF_RC"\n",
+					DP_RC(rc));
+		/* Error leaves data covered by replicas vulnerable to vos
+		 * delete, so don't advance coordination epoch.
+		 */
+			rc = 0;
+			if (cur_stripenum <
+			    agg_entry->ae_cur_stripe.as_stripenum &&
+			agg_entry->ae_cur_stripe.as_stripenum <
+			this_stripenum) {
+				/* Handle holdover stripe */
+				rc = agg_process_stripe(agg_entry);
+				if (rc)
+					D_ERROR("Holdover returned "DF_RC"\n",
+						DP_RC(rc));
+				rc = 0;
+			}
+		}
+		agg_entry->ae_cur_stripe.as_stripenum = this_stripenum;
+	}
+
+	/* Add the extent to the entry, for the current stripe */
+	D_ALLOC_PTR(extent);
+	if (extent == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
+
+	extent->ae_recx = entry->ie_recx;
+	extent->ae_orig_recx = entry->ie_orig_recx;
+	extent->ae_epoch = entry->ie_epoch;
+	agg_entry->ae_rsize = entry->ie_rsize;
+
+	d_list_add_tail(&extent->ae_link,
+			&agg_entry->ae_cur_stripe.as_dextents);
+
+	if (!agg_entry->ae_cur_stripe.as_extent_cnt)
+		/* first extent in stripe: save the start offset */
+		agg_entry->ae_cur_stripe.as_offset =  extent->ae_recx.rx_idx -
+			rounddown(extent->ae_recx.rx_idx,
+				  obj_ec_stripe_rec_nr(agg_entry->ae_oca));
+
+	agg_entry->ae_cur_stripe.as_extent_cnt++;
+	agg_entry->ae_cur_stripe.as_stripe_fill +=
+		agg_in_stripe(agg_entry, &entry->ie_recx);
+
+	if (extent->ae_epoch > agg_entry->ae_cur_stripe.as_hi_epoch)
+		agg_entry->ae_cur_stripe.as_hi_epoch = extent->ae_epoch;
+
+	D_DEBUG(DB_TRACE, "adding extent %lu,%lu, to stripe  %lu, shard: %u\n",
+		extent->ae_recx.rx_idx, extent->ae_recx.rx_nr,
+		agg_stripenum(agg_entry, extent->ae_recx.rx_idx),
+		agg_entry->ae_oid.id_shard);
+out:
+	return rc;
+}
+
+/* Post iteration call back for akey.
+ */
+static int
+agg_akey_post(daos_handle_t ih, vos_iter_entry_t *entry,
+	      struct ec_agg_entry *agg_entry, unsigned int *acts)
+{
+	int rc = 0;
+
+	if (agg_entry->ae_cur_stripe.as_extent_cnt) {
+		rc = agg_process_stripe(agg_entry);
+
+		agg_entry->ae_cur_stripe.as_stripenum	= 0UL;
+		agg_entry->ae_cur_stripe.as_hi_epoch	= 0UL;
+		agg_entry->ae_cur_stripe.as_stripe_fill = 0UL;
+		agg_entry->ae_cur_stripe.as_extent_cnt	= 0U;
+		agg_entry->ae_cur_stripe.as_offset	= 0U;
```

Comment:
need to cleanup agg_entry->ae_cur_stripe.as_dextents right? possibly one remaining last cross stripe extent left there?
or better to assert agg_entry->ae_cur_stripe.as_dextents is empty